src/m3ninx/index/segment/mem/concurrent_postings_map.go

```diff
@@ -100,23 +101,23 @@ func (m *concurrentPostingsMap) Get(key []byte) (postings.List, bool) {
 // GetRegex returns the union of the postings lists whose keys match the
 // provided regexp.
 func (m *concurrentPostingsMap) GetRegex(re *regexp.Regexp) (postings.List, bool) {
-	var pl postings.MutableList
+	lists := make([]postings.List, 0, m.postingsMap.Len())
 
 	m.RLock()
 	for _, mapEntry := range m.postingsMap.Iter() {
 		// TODO: Evaluate lock contention caused by holding on to the read lock while
 		// evaluating this predicate.
 		// TODO: Evaluate if performing a prefix match would speed up the common case.
 		if re.Match(mapEntry.Key()) {
-			// TODO: Sort the posting lists so that we take the union in order of decreasing size.
-			if pl == nil {
-				pl = mapEntry.Value().CloneAsMutable()
-			} else {
-				_ = pl.UnionInPlace(mapEntry.Value())
-			}
+			lists = append(lists, mapEntry.Value())
```

Comment:
I think since you already use `append`, it would make sense to use `nil` list and let `append` to do all allocations.

Reply:
If we leave it for `append()` to do allocations, then in cases where there are at least two matches we will end up with more allocations than doing it manually (at least from my testing, go runtime seems to allocate in the powers of two: 1, 2, 4, 8, ...). And I think we would like to avoid unnecessary allocations here.

Reply:
So I guess it depends on how many matches we get. If it's mostly 0 or 1 then `append` is better, otherwise it's better to pre-allocate. I'll leave to you to decide which outcome is most likely here.